Can't build any programs on 32-bit ARM

Linking any Crystal program (even an empty one) on an arm architecture fails with the following error:

```console
$ cc .build/crystal.o -o .build/crystal  -rdynamic -L/usr/bin/../lib/crystal src/llvm/ext/llvm_ext.o `"/usr/bin/llvm-config-12" --libs --system-libs --ldflags 2> /dev/null` -lstdc++ -lpcre -lm -lgc -lpthread -levent -lrt -ldl
/usr/bin/ld: .build/crystal.o: in function `__muloti4':
/home/thesseyren/Masaüstü/crystal-arm/crystal/src/crystal/compiler_rt/mul.cr:42: undefined reference to `__multi3'
/usr/bin/ld: .build/crystal.o: in function `//':
/home/thesseyren/Masaüstü/crystal-arm/crystal/src/int.cr:105: undefined reference to `__multi3'
/usr/bin/ld: /home/thesseyren/Masaüstü/crystal-arm/crystal/src/int.cr:105: undefined reference to `__multi3'
/usr/bin/ld: .build/crystal.o: in function `_carrying_mul':
/home/thesseyren/Masaüstü/crystal-arm/crystal/src/crystal/compiler_rt/divmod128.cr:45: undefined reference to `__multi3'
/usr/bin/ld: .build/crystal.o: in function `_carrying_mul_add':
/home/thesseyren/Masaüstü/crystal-arm/crystal/src/crystal/compiler_rt/divmod128.cr:56: undefined reference to `__multi3'
/usr/bin/ld: .build/crystal.o:/home/thesseyren/Masaüstü/crystal-arm/crystal/src/crystal/compiler_rt/divmod128.cr:161: more undefined references to `__multi3' follow
collect2: error: ld returned 1 exit status
```

The function `__multi3` is undefined in `compiler-rt` on 32-bit ARM.

There should be no implicit 128-bit arithmetics in Crystal program, unless you explicitly use those data types. Even the compiler has not yet support for parsing 128-bit integer literals.

#11206 added `__muloti4` etc. which do 128-bit arithmetics and are defined as a `fun`. So they' always get compiled, even when never called.

A quick temporary workaround is this patch, which disables the troublesome functions on arm. They're currently not necessary unless you *actually* use 128-bit arithmetics.
```diff
--- i/src/crystal/compiler_rt/divmod128.cr
+++ w/src/crystal/compiler_rt/divmod128.cr
@@ -1,4 +1,5 @@
 # This file includes an implementation of (U)Int128 modulo/division operations
+{% skip_file if flag?(:arm) %}

 # :nodoc:
 fun __divti3(a : Int128, b : Int128) : Int128
diff --git i/src/crystal/compiler_rt/mul.cr w/src/crystal/compiler_rt/mul.cr
index affe2f2f8..6f90b3346 100644
--- i/src/crystal/compiler_rt/mul.cr
+++ w/src/crystal/compiler_rt/mul.cr
@@ -39,4 +39,3 @@ end

 __mul_impl(__mulosi4, Int32, 32)
 __mul_impl(__mulodi4, Int64, 64)
-__mul_impl(__muloti4, Int128, 128)
+{% unless flag?(:arm) %}
+  __mul_impl(__muloti4, Int128, 128)
+{% end %}
```

Of course, this is not a real solution because we want to support 128-bit arithmetics on all platforms. Implementing the compiler-rt functions ourselves is actually a step to enable that for 32-bit architectures. It seems we need to implement __multi3 as well. At least for arm.

This was originally reported in https://forum.crystal-lang.org/t/cant-compile-crystal-1-2-2-on-32-bit-arm-armv7l-device-undefined-reference-to-multi3/4077/7

i386 seems to be unaffected, so apparently `__multi3` is available.

--- a/spec/std/crystal/compiler_rt/multi3_spec.cr
+++ b/spec/std/crystal/compiler_rt/multi3_spec.cr
@@ -3,17 +3,6 @@ require "crystal/compiler_rt/multi3"
 
 # Ported from https://github.com/llvm/llvm-project/blob/ce59ccd04023cab3a837da14079ca2dcbfebb70c/compiler-rt/test/builtins/Unit/multi3_test.c
 
-private def test__mulodi4(a : Int64, b : Int64, expected : Int64, expected_overflow : Int32, file = __FILE__, line = __LINE__)
-  it "passes compiler-rt builtins unit tests" do
-    actual_overflow : Int32 = 0
-    actual = __mulodi4(a, b, pointerof(actual_overflow))
-    actual_overflow.should eq(expected_overflow), file: file, line: line
-    if !expected_overflow
-      actual.should eq(expected), file: file, line: line
-    end
-  end
-end
-
 # TODO: Replace helper methods with literals once possible
 
 private def make_ti(a : Int128, b : Int128)

--- a/spec/std/crystal/compiler_rt/multi3_spec.cr
+++ b/spec/std/crystal/compiler_rt/multi3_spec.cr
@@ -1,0 +1,63 @@
+require "spec"
+require "crystal/compiler_rt/multi3"
+
+# Ported from https://github.com/llvm/llvm-project/blob/ce59ccd04023cab3a837da14079ca2dcbfebb70c/compiler-rt/test/builtins/Unit/multi3_test.c
+
+private def test__mulodi4(a : Int64, b : Int64, expected : Int64, expected_overflow : Int32, file = __FILE__, line = __LINE__)
+  it "passes compiler-rt builtins unit tests" do
+    actual_overflow : Int32 = 0
+    actual = __mulodi4(a, b, pointerof(actual_overflow))
+    actual_overflow.should eq(expected_overflow), file: file, line: line
+    if !expected_overflow
+      actual.should eq(expected), file: file, line: line
+    end
+  end
+end
+
+# TODO: Replace helper methods with literals once possible
+
+private def make_ti(a : Int128, b : Int128)
+  (a << 64) + b
+end
+
+it ".__multi3" do
+  __multi3(0, 0).should eq 0
+  __multi3(0, 1).should eq 0
+  __multi3(1, 0).should eq 0
+  __multi3(0, 10).should eq 0
+  __multi3(10, 0).should eq 0
+  __multi3(0, 81985529216486895).should eq 0
+  __multi3(81985529216486895, 0).should eq 0
+  __multi3(0, -1).should eq 0
+  __multi3(-1, 0).should eq 0
+  __multi3(0, -10).should eq 0
+  __multi3(-10, 0).should eq 0
+  __multi3(0, -81985529216486895).should eq 0
+  __multi3(-81985529216486895, 0).should eq 0
+  __multi3(1, 1).should eq 1
+  __multi3(1, 10).should eq 10
+  __multi3(10, 1).should eq 10
+  __multi3(1, 81985529216486895).should eq 81985529216486895
+  __multi3(81985529216486895, 1).should eq 81985529216486895
+  __multi3(1, -1).should eq -1
+  __multi3(1, -10).should eq -10
+  __multi3(-10, 1).should eq -10
+  __multi3(1, -81985529216486895).should eq -81985529216486895
+  __multi3(-81985529216486895, 1).should eq -81985529216486895
+  __multi3(3037000499, 3037000499).should eq 9223372030926249001
+  __multi3(-3037000499, 3037000499).should eq -9223372030926249001
+  __multi3(3037000499, -3037000499).should eq -9223372030926249001
+  __multi3(-3037000499, -3037000499).should eq 9223372030926249001
+  __multi3(4398046511103, 2097152).should eq 9223372036852678656
+  __multi3(-4398046511103, 2097152).should eq -9223372036852678656
+  __multi3(4398046511103, -2097152).should eq -9223372036852678656
+  __multi3(-4398046511103, -2097152).should eq 9223372036852678656
+  __multi3(2097152, 4398046511103).should eq 9223372036852678656
+  __multi3(-2097152, 4398046511103).should eq -9223372036852678656
+  __multi3(2097152, -4398046511103).should eq -9223372036852678656
+  __multi3(-2097152, -4398046511103).should eq 9223372036852678656
+  __multi3(
+    make_ti(0x00000000000000B5, 0x04F333F9DE5BE000),
+    make_ti(0x0000000000000000, 0x00B504F333F9DE5B)
+  ).should eq make_ti(0x7FFFFFFFFFFFF328, 0xDF915DA296E8A000)
+end

--- a/src/crystal/compiler_rt.cr
+++ b/src/crystal/compiler_rt.cr
@@ -2,3 +2,8 @@
 
 require "./compiler_rt/mul.cr"
 require "./compiler_rt/divmod128.cr"
+
+{% if flag?(:arm) %}
+  # __multi3 was only missing on arm
+  require "./compiler_rt/multi3.cr"
+{% end %}

--- a/src/crystal/compiler_rt/multi3.cr
+++ b/src/crystal/compiler_rt/multi3.cr
@@ -1,0 +1,32 @@
+private def __mulddi3(a : UInt64, b : UInt64) : Int128
+  bits_in_dword_2 = (sizeof(Int64) &* 8) // 2
+  lower_mask = ~0_u64 >> bits_in_dword_2
+
+  low = (a & lower_mask) &* (b & lower_mask)
+  t = low >> bits_in_dword_2
+  low &= lower_mask
+  t &+= (a >> bits_in_dword_2) &* (b & lower_mask)
+  low &+= (t & lower_mask) << bits_in_dword_2
+  high = t >> bits_in_dword_2
+  t = low >> bits_in_dword_2
+  low &= lower_mask
+  t &+= (b >> bits_in_dword_2) &* (a & lower_mask)
+  low &+= (t & lower_mask) << bits_in_dword_2
+  high &+= t >> bits_in_dword_2
+  high &+= (a >> bits_in_dword_2) &* (b >> bits_in_dword_2)
+
+  (high.to_i128! << 64) &+ low
+end
+
+# :nodoc:
+# Ported from https://github.com/llvm/llvm-project/blob/ce59ccd04023cab3a837da14079ca2dcbfebb70c/compiler-rt/lib/builtins/multi3.c
+fun __multi3(a : Int128, b : Int128) : Int128
+  a_low = (a & ~0_u64).to_u64!
+  a_high = (a >> 64).to_i64!
+  b_low = (b & ~0_u64).to_u64!
+  b_high = (b >> 64).to_i64!
+
+  result = __mulddi3(a_low, b_low)
+  result &+= ((a_high &* b_low) &+ (a_low &* b_high)).to_i128! << 64
+  result
+end

